memiavl/db.go

```diff
@@ -184,43 +215,71 @@ func (db *DB) checkBackgroundSnapshotRewrite() error {
 			return fmt.Errorf("background snapshot rewriting failed: %w", result.err)
 		}
 
-		// snapshot rewrite succeeded, catchup and switch
-		if err := result.mtree.CatchupWAL(db.wal, 0); err != nil {
-			return fmt.Errorf("catchup failed: %w", err)
-		}
-		if err := db.reloadMultiTree(result.mtree); err != nil {
-			return fmt.Errorf("switch multitree failed: %w", err)
+		db.pendingForSwitch = result.mtree
+	default:
+	}
+
+	if db.pendingForSwitch == nil {
+		return nil
+	}
+
+	// catchup the remaining wal
+	if err := db.pendingForSwitch.CatchupWAL(db.wal, 0); err != nil {
+		return fmt.Errorf("catchup failed: %w", err)
+	}
+
+	// make sure there are at least a few queryable states after switch
+	if db.pendingForSwitch.Version() < db.pendingForSwitch.SnapshotVersion()+int64(db.minQueryStates) {
+		return nil
+	}
+
+	// do the switch
+	if err := db.reloadMultiTree(db.pendingForSwitch); err != nil {
+		return fmt.Errorf("switch multitree failed: %w", err)
+	}
+	db.logger.Info("switched to new snapshot", "version", db.MultiTree.Version())
+
+	db.pendingForSwitch = nil
+	db.pruneSnapshots(db.SnapshotVersion())
+	return nil
+}
+
+// pruneSnapshot prune the old snapshots
+func (db *DB) pruneSnapshots(newSnapshotVersion int64) {
+	// wait until last prune finish
+	db.pruneSnapshotLock.Lock()
+
+	go func() {
+		defer db.pruneSnapshotLock.Unlock()
+
+		entries, err := os.ReadDir(db.dir)
+		if err != nil {
+			db.logger.Error("failed to read db dir", "err", err)
+			return
 		}
-		// prune the old snapshots
-		// wait until last prune finish
-		db.pruneSnapshotLock.Lock()
-		go func() {
-			defer db.pruneSnapshotLock.Unlock()
-
-			entries, err := os.ReadDir(db.dir)
-			if err == nil {
-				for _, entry := range entries {
-					if entry.IsDir() && strings.HasPrefix(entry.Name(), SnapshotPrefix) {
-						currentVersion, err := strconv.ParseInt(strings.TrimPrefix(entry.Name(), SnapshotPrefix), 10, 32)
-						if err != nil {
-							fmt.Printf("failed when parse current version: %s\n", err)
-							continue
-						}
-						if result.mtree.metadata.CommitInfo.Version-currentVersion > int64(db.snapshotKeepRecent) {
-							fullPath := filepath.Join(db.dir, entry.Name())
-							if err := os.RemoveAll(fullPath); err != nil {
-								fmt.Printf("failed when remove old snapshot: %s\n", err)
-							}
-						}
+
+		for _, entry := range entries {
+			if entry.IsDir() && strings.HasPrefix(entry.Name(), SnapshotPrefix) {
+				snapshotVersion, err := strconv.ParseInt(strings.TrimPrefix(entry.Name(), SnapshotPrefix), 10, 32)
+				if err != nil {
+					db.logger.Error("failed when parse snapshot file name", "err", err)
+					continue
+				}
+				if snapshotVersion < newSnapshotVersion-int64(db.snapshotKeepRecent) {
```

Comment:
should be `snapshotKeepRecent* snapshotInterval` ?

Reply:
fixed, changed to count the snapshot numbers, instead of version range.